### PR TITLE
[SYCL][CUDA] Windows and MSVC support for CUDA backend

### PR DESCRIPTION
**Pull request by AidanBeltonS**
Patch adds windows support for CUDA backend. 

Adds general handling of Windows file paths
Windows support is enabled with remangling of variables from PR #4207 as it fixes mismatch between windows 32-bit longs and default 64-bit long and handles wchar_size.
Adds changes to account for MSVC's default to private classes.
Fixes to unittests for windows.

### CODE REVIEW

File: libclc/utils/prepare-builtins.cpp [hunk 1]
```diff
@@ -75,6 +75,17 @@ int main(int argc, char **argv) {
   if (NamedMDNode *OCLVersion = M->getNamedMetadata("opencl.ocl.version"))
       M->eraseNamedMetadata(OCLVersion);
 
+  // Drop wchar_size module flag
```
Comment (bader) [inline libclc/utils/prepare-builtins.cpp:78]: Could you add clarification why we drop this flag, please? See a good example in the comments for the previous if-statement.

**Review by bader — DISMISSED** [review]
`libclc` changes look good to me. Just one minor suggestion.

File: libclc/utils/prepare-builtins.cpp (hunk 1 above)
Comment (AidanBeltonS) [reply libclc/utils/prepare-builtins.cpp:78]: Sure, would this be okay. FYI @steffenlarsen 

 // Drop wchar_size module flag. Windows adds the wchar_size variable 
  // which is not used in linux. libclc inherits the wchar_size flags 
  // from windows so this must be dropped before mangling. The wchar_size
  // module flag is taken out by the non-wchar_size flags are added back 
  // into the module.

Comment (steffenlarsen) [reply libclc/utils/prepare-builtins.cpp:78]: I think you could simplify it to:

> wchar_size flag can cause a mismatch between libclc libraries and modules using them. Since wchar is not used by libclc we drop the flag.

Comment (AidanBeltonS) [reply libclc/utils/prepare-builtins.cpp:78]: Okay sounds good

File: clang/lib/Driver/ToolChains/Cuda.cpp [hunk 2]
```diff
@@ -722,8 +735,14 @@ void CudaToolChain::addClangTargetOptions(
       llvm::sys::path::append(WithInstallPath, Twine("../../../share/clc"));
       LibraryPaths.emplace_back(WithInstallPath.c_str());
 
+      // Select remangled libclc variant. 64-bit longs default, 32-bit longs on
+      // Windows
       std::string LibSpirvTargetName =
           "remangled-l64-signed_char.libspirv-nvptx64--nvidiacl.bc";
+      if (HostTC.getTriple().isOSWindows())
+        LibSpirvTargetName =
+            "remangled-l32-signed_char.libspirv-nvptx64--nvidiacl.bc";
```
Comment (mdtoguchi) [inline clang/lib/Driver/ToolChains/Cuda.cpp:744]: There doesn't seem to be any tests checking that the proper `.bc` file is being picked up.  Is it necessary?

**Comment by bader** [comment]
@AidanBeltonS, this change introduces a regression: http://ci.llvm.intel.com:8010/#/builders/37/builds/11953/steps/14/logs/FAIL__Clang__cuda-windows_cu. Could you take a look, please?

**Comment by AidanBeltonS** [comment]
> @AidanBeltonS, this change introduces a regression: http://ci.llvm.intel.com:8010/#/builders/37/builds/11953/steps/14/logs/FAIL__Clang__cuda-windows_cu. Could you take a look, please?

Yes, I can try and see what is causing it. Thanks

File: clang/lib/Driver/ToolChains/Cuda.cpp (hunk 2 above)
Comment (AidanBeltonS) [reply clang/lib/Driver/ToolChains/Cuda.cpp:744]: That is a good idea. Ill write a test to check that the correct file is being used.

Comment (AidanBeltonS) [reply clang/lib/Driver/ToolChains/Cuda.cpp:746]: Added test to check .bc file: clang/test/Driver/cuda-nvptx-target.cpp

**Comment by AidanBeltonS** [comment]
> @AidanBeltonS, this change introduces a regression: http://ci.llvm.intel.com:8010/#/builders/37/builds/11953/steps/14/logs/FAIL__Clang__cuda-windows_cu. Could you take a look, please?

Added a fix for the cuda-windows.cu test. Hopefully it should run without error.

**Review by bader — DISMISSED** [review]

**Review by mdtoguchi — APPROVED** [review]

**Review by steffenlarsen — APPROVED** [review]
I've built and tested it on a Windows machine using MSVC and only 2 tests in the test suite (https://github.com/intel/llvm-test-suite/pull/383 specifically) fail. I suggest we disregard those 2 tests until both this and https://github.com/intel/llvm-test-suite/pull/383 is merged, then mark them as failing for the CUDA backend on Windows and make issues for each failing test.

Additions LGTM!

**Comment by bader** [comment]
@smaslov-intel, @intel/llvm-reviewers-runtime, ping.

@AidanBeltonS, could you resolve merge conflict, please?

**Comment by AidanBeltonS** [comment]
> @AidanBeltonS, could you resolve merge conflict, please?

Conflict should be resolved now.

File: sycl/plugins/cuda/CMakeLists.txt [hunk 3]
```diff
@@ -37,6 +45,24 @@ target_link_libraries(pi_cuda
...
+
+  # This script file is used to allow exporting pi* symbols only.
+  # All other symbols are regarded as local (hidden)
+  set(linker_script "${CMAKE_CURRENT_SOURCE_DIR}/../ld-version-script.txt")
+
+  # Filter symbols based on the scope defined in the script file,
+  # and export pi* function symbols in the library.
+  target_link_libraries(pi_cuda PRIVATE "-Wl,--version-script=${linker_script}")
+endif()
+
```
Comment (smaslov-intel) [inline sycl/plugins/cuda/CMakeLists.txt:65]: NIT: this code is duplicated in each plugin's CMakeLists.txt. Can't we move it to sycl/plugins/CMakeLists.txt

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 4]
```diff
@@ -4725,7 +4725,11 @@ pi_result cuda_piextUSMGetMemAllocInfo(pi_context context, const void *ptr,
                        PI_MEM_TYPE_HOST);
       }
       // should never get here
+#ifdef _MSC_VER
+      __assume(0);
+#else
       __builtin_unreachable();
+#endif
```
Comment (smaslov-intel) [inline sycl/plugins/cuda/pi_cuda.cpp:4732]: this deserves to be in an utility function

File: sycl/plugins/level_zero/CMakeLists.txt [hunk 5]
```diff
@@ -58,6 +58,15 @@ if (NOT DEFINED LEVEL_ZERO_LIBRARY OR NOT DEFINED LEVEL_ZERO_INCLUDE_DIR)
...
 
+  if (WIN32)
+    # Copy DLL into binary directory
+    ExternalProject_Add_Step(level-zero-loader llvmbininstall
+      COMMAND ${CMAKE_COMMAND} -E copy_directory <INSTALL_DIR>/bin/ ${LLVM_BINARY_DIR}/bin
+      COMMENT "Installing level-zero-loader into the LLVM binary directory"
+      DEPENDEES install
+    )
+  endif()
+
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/CMakeLists.txt:69]: why is this change necessary in this PR and in general?

File: sycl/unittests/pi/cuda/test_commands.cpp [hunk 6]
```diff
@@ -30,7 +30,7 @@ struct CudaCommandsTest : public ::testing::Test {
 
   void SetUp() override {
     // skip the tests if the CUDA backend is not available
-    if (!plugin) {
+    if (!plugin.has_value()) {
```
Comment (smaslov-intel) [inline sycl/unittests/pi/cuda/test_commands.cpp:33]: why do you need .has_value()? there is a conversion to bool defined: https://en.cppreference.com/w/cpp/utility/optional/operator_bool

**Review by smaslov-intel — COMMENTED** [review]
Please see comments

File: sycl/plugins/level_zero/CMakeLists.txt (hunk 5 above)
Comment (steffenlarsen) [reply sycl/plugins/level_zero/CMakeLists.txt:69]: On Windows the .dll for the level-zero loader was in /bin but wasn't copied to the install folder. This would cause testing to fail as it insisted on loading it. Arguably it could be a separate PR.

File: sycl/unittests/pi/cuda/test_commands.cpp (hunk 6 above)
Comment (steffenlarsen) [reply sycl/unittests/pi/cuda/test_commands.cpp:33]: Frankly it was just a personal preference when using `std::optional`. @AidanBeltonS Feel free to scrap the `.has_value()`.

File: sycl/plugins/level_zero/CMakeLists.txt (hunk 5 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/CMakeLists.txt:69]: Let's keep it separate. Today Level-zero testing is working, I assume, so I don't understand this fix at least. Aren't DLL searched in %PATH% on Windows?

File: sycl/unittests/pi/cuda/test_commands.cpp (hunk 6 above)
Comment (smaslov-intel) [reply sycl/unittests/pi/cuda/test_commands.cpp:33]: This is a NIT, feel free to leave it as is.

File: sycl/plugins/level_zero/CMakeLists.txt (hunk 5 above)
Comment (steffenlarsen) [reply sycl/plugins/level_zero/CMakeLists.txt:69]: > Aren't DLL searched in %PATH% on Windows?

I believe so, but the problem I encountered while testing was that the .dll for the level-zero loader wasn't copied into anywhere the tests could find it with the environment they set up. I tried rebuilding without these changes and it seems to have been resolved separately.

@AidanBeltonS These changes to sycl/plugins/level_zero/CMakeLists.txt can be removed.

**Comment by steffenlarsen** [comment]
A new CUDA backend unittest has been added in sycl\unittests\pi\cuda\test_contexts.cpp. This will need the same treatment as the other unittests.

File: sycl/plugins/cuda/pi_cuda.cpp (hunk 4 above)
Comment (AidanBeltonS) [reply sycl/plugins/cuda/pi_cuda.cpp:4732]: Do you have a location in mind for this utility function? As this function would not want to be relegated just to pi_cuda.cpp, it would not make sense to place this utility func in pi_cuda.h
Is there a location in dpc++ for these sorts of functions?

File: sycl/plugins/cuda/CMakeLists.txt (hunk 3 above)
Comment (AidanBeltonS) [reply sycl/plugins/cuda/CMakeLists.txt:65]: Yes, it can be moved to sycl/plugins/CMakeLists.txt
I will make the change

**Review by bader — APPROVED** [review]

File: sycl/plugins/cuda/CMakeLists.txt (hunk 3 above)
Comment (AidanBeltonS) [reply sycl/plugins/cuda/CMakeLists.txt:65]: This change to the CMake was causing a build error. Probably due to operations like add_library, add_dependencies, set_target_properties, target_include_directories, and target_link_libraries not being executed yet. This operations are not the same across each plugin so can't be wrapped up in sycl/plugins/CMakeLists.txt easily. So, it may be more verbose than necessary but the keeping this code in the each plugin's CMakeList.txt is probably the simplest approach.

Comment (smaslov-intel) [reply sycl/plugins/cuda/CMakeLists.txt:65]: I am OK with that.

File: sycl/plugins/cuda/pi_cuda.cpp (hunk 4 above)
Comment (smaslov-intel) [reply sycl/plugins/cuda/pi_cuda.cpp:4732]: I don't think we want plugins have dependency on the runtime, but we could create shared utilities for all plugins (many existing handling may be unified across plugins too). I think this deserves a separate PR and this can go in as is.

**Review by bader — APPROVED** [review]

**Review by steffenlarsen — APPROVED** [review]

**Review by mdtoguchi — APPROVED** [review]
